[Feat] 고민 작성시 고민 상세 이동 로직 구현

daminoworld opened the pull request:
## 💪 작업한 내용
- 기존에 작성탭을 눌렀을때 writeVC 자체를 present 하는 대신 writeVC를 rootVC로 하는 BaseNC를 만들어 navigationController 자체를 present하여 띄워주도록 변경
- 탭바 컨트롤러에서 작성탭에서 writeVC를 루트뷰로 하는 NC를 present하였기 때문에 탭바 -> NC(writeVC) -> pickerVC 구조에서 self인 pickerVC를 present하는 presentingViewController는 NC이다.
- 때문에 writeNC라는 이름으로 self.presentingViewController를 바인딩하고, 이때는 타입이 viewController 이기 때문에 NC라는 이름으로 다시 BaseNC로 타입캐스팅을 해주어 BaseNC 타입의 인스턴스 NC를 구한다.
- 그리고 리스폰스로 받은 데이터를 이용해 worryDetailVC의 인스턴스를 생성하고, NC의 rootVC를 worryDetailVC로 바꾸어 popToRootViewController를 이용해 고민상세 뷰로 화면전환 한다.
- 이를 통해 기존에 띄워지는 고민 작성 뷰를 화면에 노출시키지 않고 바로 고민상세로 가고, 고민상세를 닫을때 바로 기존 탭을 보여질 수 있다.

## 💜 PR Point
<!-- 피드백을 받고 싶은 부분, 공유하고 싶은 부분, 작업 과정, 이유를 적어주세요. -->
- 이번 이슈를 처리하는 생각보다 시간이 꽤 걸렸습니다...

기존 구현은 작성탭 클릭시 writeVC에서 부터 각 vc들을 present 방식으로 보여주는데 

홈 -> 고민 작성 -> 데드라인 과 같은 플로우에서 데드라인 창을 dismiss하고  고민 상세를 보여줘야하는데 
이를 구현하기 위해서 처음에 2가지 방법을 생각했는데 
1. 홈 -> 고민 작성 -> 데드라인 -> 데드라인 dismiss -> 고민작성 dismiss -> 홈 -> 고민 상세 present
2.  홈 -> 고민작성 -> 데드라인 -> 데드라인 dismiss -> 고민 상세 present 

1번의 경우 고민 상세에서 창을 닫을 때 바로 홈으로 갈 수 있지만 고민 상세로 가는 과정에서 홈화면이 보여지기 때문에 어색하다는 문제가 있고
2번의 경우 고민상세는 바로 자연스럽게 띄울 수 있는데 고민 상세뷰가 고민 작성 창에서 띄운것이라 고민 상세를 내릴때 고민 작성창이 보인다는 문제가 있었습니다.

| 1번 홈이 보임 | 2번 작성이 보임 |
| ---- | ---- | 
|  ![Simulator Screen Recording - iPhone 13 mini - 2023-12-27 at 16 05 21](https://github.com/TeamHARA/KAERA_iOS/assets/32871014/10998adb-30e4-48f6-a406-6a7360d1a8f7) | ![Simulator Screen Recording - iPhone 13 mini - 2023-12-27 at 15 57 53](https://github.com/TeamHARA/KAERA_iOS/assets/32871014/4843ec3c-2d61-4488-8fb5-6a47be9176a5) |

present 방식이라 플로우 중간의 화면을 바꿔치기 할 수 없어서 transition을 바꿔보고 present 및 dismiss 순서를 조금씩 조정해보면서 자연스럽게 화면전환이 될 수 있도록 1,2번 방식에서 수정해보려 했으나 화면 전환 플로우 상에 홈이나 고민작성이 있기 때문에 1,2번 방식으로는 해결이 어려웠습니다.

- 그래서 결국 화면을 스택처럼 쌓아서 화면전환 시켜주는 NavigationController를 사용해 rootVC를 바꿔주는 방식으로 문제를 해결하였습니다.
  - writeVC 대신 writeVC를 rootVC로 하는 NC자체를 탭바에서 present하고 
  - 고민 완료시 NC에서 setViewControllers를 이용해 NC의 rootVC를 worryDetailVC로 바꾸고 rootVC 즉, 고민 상세로 이동하도록 하였습니다.

## 📸 스크린샷
<!-- gif or mp4 용량 제한이 있는데... 용량 넘어가면 슬랙으로 보내 주세요. -->
> NavigationController를 사용

![Simulator Screen Recording - iPhone 13 mini - 2023-12-27 at 15 54 50](https://github.com/TeamHARA/KAERA_iOS/assets/32871014/9392629d-5129-47ad-8fe2-7b0efbc3db6f)


## 🚨 관련 이슈
- Resolved: #141


<!-- 아 맞다! Assignee, Reviewer 설정! 😇 -->

daminoworld committed:
Chore: #141-Remove comment

daminoworld committed:
Chore: #141-Remove duplicated code

- modalPresentationstyle 코드의 중복으로 .overCurrentContext부분 삭제

daminoworld committed:
Chore: #141-Set explicitly sequence of transition

- 고민 캐기 뷰가 dismiss되고 상세뷰가 dismiss될 수 있도록 decisionVC의 completion에 detailVC의 dismiss가 실행되도록 변경

daminoworld committed:
Feat: #141-Present NC instead VC

- 기존에 작성탭을 눌렀을때 writeVC 자체를 present 하는 대신 writeVC를 rootVC로 하는 BaseNC를 만들어 navigationController 자체를 present하여 띄워주도록 변경

daminoworld committed:
Feat: #141-Switch rootVC of NC with worryDetailVC

- 탭바 컨트롤러에서 작성탭에서 writeVC를 루트뷰로 하는 NC를 present하였기 때문에 현재 pickerVC self를 present하는 presentingViewController는 BaseNC이다.
- 때문에 writeNC라는 이름으로 self.presentingViewController를 바인딩하고, 이를 NC라는 이름으로 BaseNC로 형변환 해주어 BaseNC 인스턴스 NC를 구한다.
- 그리고 리스폰스로 받은 데이터를 이용해 worryDetailVC의 인스턴스를 생성하고, NC의 rootVC를 worryDetailVC로 바꾸어 popToRootViewController를 이용해 고민상세 뷰로 화면전환 한다.
- 이를 통해 기존에 띄워지는 고민 작성 뷰를 화면에 노출시키지 않고 바로 고민상세로 가고, 고민상세를 닫을때 바로 기존 탭을 보여질 수 있다.

oy6uns commented on KAERA/KAERA/Scenes/Writing/View/WritePickerVC.swift:
```diff
@@ -132,18 +132,20 @@ class WritePickerVC: BaseVC {
-                    // TODO: 고민 작성 후 해당 고민 상세로 이동
                     self?.dismiss(animated: false) {
-                        writeVC.dismiss(animated: true)
+                        let worryDetailVC = HomeWorryDetailVC(worryId: data.worryId, type: .digging)
+                        NC.setViewControllers([worryDetailVC], animated: true)
+                        NC.popToRootViewController(animated: true)
```
NC.setViewControllers([worryDetailVC], animated: true)
코드에서 이미 navigationController의 스택을 worryDetailVC하나로 설정해서, 아래의 popToRootViewController가 없더라도 코드가 정상적으로 돌아가지 않을까하는 생각이 들었어요!

```suggestion
                        let worryDetailVC = HomeWorryDetailVC(worryId: data.worryId, type: .digging)
                        NC.setViewControllers([worryDetailVC], animated: true)
                        /// NC.popToRootViewController(animated: true)
```

위처럼 마지막 줄을 제거하면 문제가 발생할까요?

daminoworld replied:
오호 제가 생각했던거는 setViewContollers로 navigationControllers의 viewController들을 설정하고 popToRootViewController 메서드로 이동하는거라고 생각했었는데
공식문서를 보니 setViewContollers 실행시 설정된 top viewController로 이동까지 시켜주네요 popToRootVC 부분 삭제해서 실행해서 확인해보니 잘 작동합니다 👍 감사합니다
 popToRootViewController부분은 삭제하였습니다!

oy6uns reviewed (approved):
아예 tabbarController 코드를 수정해서 NavigationController를 modal방식으로 띄우는 생각을 한거 대박이네용 역시 최고👍🏻 코드리뷰 하나 달아놓은 것만 확인해주세용~~

고민의 흔적이 보였던 풀리퀘였던 것 같습니다. 고생많으셨습니다.

daminoworld committed:
Fix: #141-Remove popToRootViewController method

- NC.setViewControllers로 worryDetailVC로 바로 이동하기 때문에 popTooRootViewController는 필요없어 삭제함
